README.md
# python-app-setup
A boilerplate setup for basic stuff like linting, prettier, changelog, tests for a python app

### Run locally

- clone the repo `git clone git@github.com:dbads/python-app-setup.git`
- create a virtual environment `python3 -m venv path/to/venv`
- install the requrirements `pip install -r requirements.txt`
- install the recommended  vscode extensions if using the vscode or similar extensions in your IDE

### Seting up linters/fixers, format on save settings

- to use flake8/pylint as a linter, enable them in `.vscode/settings.json` or globally in `settings.json`
- set a formatting provider in `settings.json`
- `settings.json` contains basic setup, see for example. You can modify the settings as per the need and use.

For any queries reach out to me at <a href="mailto:deepakbharti@mnnit.ac.in"> deepakbharti@mnnit.ac.in</a>.


CHANGELOG.md
<!-- Either manually add changes while releasing to prod of create some script to automate moving unreleased section to released -->
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project tries to adhere to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

<!-- NOTE: Don't change the formatting of this file because parsing of it depends on it. The unreleased section is automatically moved to a release on main tags -->

## [Unreleased]

> NOTE: ⬇⬇ Add below this ⬇⬇

### Changed

- Testing changelog

> NOTE: ⬆⬆ And above this ⬆⬆

---

## Deploy Notes

If anything special needs to be done for a release (like update private config) or some precautions need to be taken. Please add them below.

The unreleased ones will be moved to it's own subheading after a release automatically by the [versionCheck.js](./.gitlab/scripts/versionCheck.js) script.

### [Unreleased]

> NOTE: ⬇⬇ Add below this ⬇⬇
> NOTE: ⬆⬆ And above this ⬆⬆

---


test.py
# Notes
# 1. the meeting system is designed to handle operatinos within a single day 
# 2. in a day time is measured using 24 hour clock 1,2,3...24
# 3. time inputs are allowed only as [1,2 .... 23, 24] discrete values


class MeetingSystem:
  """ A model to create a meeting system with following methods
  - book: input format = book employee_id, start_time, end_time
  - cancel: input format = cancel employee_id, meeting_id

  Helper/utitlity methods
  - get_time_key: returns a string key for given time interval
  - confirm_room: books a room with gicen details
  """
  schedules = {
  # sample entry
  # '12': [5,7], i.e room_id 5 and 7 is busy during time 1 to 2
  }

  meetings = {
    # An sample entry
    # meeting_id: {
    #   'employee_id': demo_employee_id,
    #   'start_time': demo_start_time,
    #   'end_time': demo_end_time,
    #   'room_id': demo_room_id,
    # }
  }

  rooms_count = 0
  employees_count = 0

  # Initializing meetings system
  def __init__(self, rooms_count, employees_count):
    """ Just initiallizes the meeting system with employees_count and rooms_count
    """
    self.rooms_count = rooms_count
    self.employees_count = employees_count

  
  # Utitlity methods 
  def validate_meeting_details(self, details):
    """Checks if meeting details are valid
    - time should be in [1,24]
    - employee_id should be in [1,N], N is employees_count
    - meeting_id should exist in meetings
    - @return Boolean(True/False)
    """
    is_valid = True

    # time should be a discrete value in [1,24], end_time>start_time
    if 'start_time' in details and 'end_time' in details:
      start_time = details['start_time']
      end_time = details['end_time']
      if not type(start_time) == int or not type(end_time) == int or\
        start_time not in range(1 ,25) or end_time not in range(1, 25):
        print('Invalid time provided.')
        is_valid = False

      if start_time >= end_time:
        print('start_time should be strictly less than end_time.')
        is_valid = False
      
      if end_time - start_time > 3:
        print('Meeting should not exceed 3 hours.')
        is_valid = False

    # check if employee_id is valid
    if 'employee_id' in details:
      employee_id = details['employee_id']
      if type(employee_id) != int or employee_id not in range(1, self.employees_count + 1):
        print('Invalid Employee ID.')
        is_valid = False
    
    # validate meeting_id
    if 'meeting_id' in details:
      meeting_id = details['meeting_id']
      if meeting_id not in self.meetings:
        print('Invalid meeting_id.')
        is_valid = False
    
    if not is_valid: return False
    return True


  def get_time_key(self, start_time, end_time):
    """ Helper to get string key of time interval
    - @input start_time and end_time
    - @returns string key - for this time interval
    """
    return str(start_time) + '-' + str(end_time)


  def confirm_room(self, employee_id, start_time, end_time):
    """Schedule a meeting in the first room which is free at the requested time slot
    - @input employee_id, start_time, end_time
    - @returns void
    """
    time_key = self.get_time_key(start_time, end_time)

    # schedule a meeting in free_room_id
    # making room busy for this time interval
    if time_key not in self.schedules:
      free_room_id = 1
      self.schedules[time_key] = [free_room_id] # no room engaged this time, simple take room_id 1 and schedule meeting there
    else:
      engaged_rooms = self.schedules[time_key]
      free_room_id = len(engaged_rooms) + 1
      self.schedules[time_key].append(free_room_id)

    # create a meeting at this time
    self.meetings[len(self.meetings) + 1] = {
      'room_id': free_room_id,
      'employee_id': employee_id,
      'start_time': start_time,
      'end_time': end_time,
    }
    print('Successfuly booked room for meeting')


  def is_overlaping_time(self, start_time, end_time, new_start_time, new_end_time):
    """Checks whether time intervals overlap or not
    """
    return new_start_time in range(start_time, end_time+1) or new_end_time in range(start_time, end_time+1)


  # core methods
  def book_room(self, employee_id, start_time, end_time):
    """Schedules a meeting in a room with requested details
    """
    is_valid_details = self.validate_meeting_details({'employee_id': employee_id, 'start_time': start_time, 'end_time': end_time})

    if is_valid_details:
      time_key = self.get_time_key(start_time, end_time)

      # see if the employee has already a meeting at this time
      already_booked_meeting = False # suppose not booked a meeting already by this member
      # see if one of the room is having a meeting scheduled by employee_id
      for room_id, meeting in self.meetings.items():
        if meeting['employee_id'] == employee_id\
          and self.is_overlaping_time(meeting['start_time'], meeting['end_time'], start_time, end_time):
          already_booked_meeting = True # found a meeting booked by this employee_id at same time
          print('Employee has already a meeting scheduled at same time')

      if not already_booked_meeting:
        # even though this employee has not booked a meeting at this time, there are already other meetings going on at this time
        # see if there is a free room to schedule this meeting at this time
        engaged_rooms = self.schedules[time_key] if time_key in self.schedules else []
        if len(engaged_rooms) + 1 > self.rooms_count: # all rooms are already engaged for this time
          print('All rooms busy for the given time interval')
        else:
          self.confirm_room(employee_id, start_time, end_time)


  def cancel_room(self, employee_id, meeting_id):
    # validate input
    is_valid_details = self.validate_meeting_details({'employee_id': employee_id, 'meeting_id': meeting_id})

    if is_valid_details:
      # cancel the meeting if employee_id has created this meeting
      if self.meetings[meeting_id]['employee_id'] == employee_id:
        start_time = self.meetings[meeting_id]['start_time']
        end_time = self.meetings[meeting_id]['end_time']
        time_key = self.get_time_key(start_time, end_time)
        room_id = self.meetings[meeting_id]['room_id']
        self.schedules[time_key].remove(room_id) # free the room from schedules 
        del self.meetings[meeting_id] # remove meeting detail from meetings
        print('Successfuly canceled the meeting')
      else: print('You are not the organizer of this meeting')


def CreateMeetingSystem():
  valid_meeting_initializer = False
  while not valid_meeting_initializer:
    employees_count = input('Please Enter the no of employees > ')
    rooms_count = input('Please Enter the no of rooms > ')

    try:
      employees_count = int(employees_count)
      rooms_count = int(rooms_count)
      valid_meeting_initializer = True # valid a meeting initiallizers rooms_count and employees_count, stop calling initializer again now

      meeting_system = MeetingSystem(rooms_count, employees_count)
      print('A MeetingSystem with %s Employees and %s meeting rooms has been created.\n'%(employees_count, rooms_count))
      
      print('Notes :-\n')
      print('1. Employee IDs should be in [1-N], N is the employee_count')
      print('2. Time should be in [1,24]\n')

      request_type = 1
      while request_type != 'quit':
        request_type = input('Enter the request_type [book, cancel, quit] > ')
        if request_type not in ['book', 'cancel', 'quit']:
          print('Operations not allowed. Please try again ...')
        
        # processing book Operations
        if request_type == 'book':
          try:
            employee_id = input('Employee ID > ').split(' ')[0]
            start_time = input('Start time > ').split(' ')[0]
            end_time = input('End time > ').split(' ')[0]
            meeting_system.book_room(int(employee_id), int(start_time), int(end_time))
          except Exception as e:
            print('Error while booking room. try again ...', e)
            print(
                  type(e).__name__,          # TypeError
                  __file__,                  # /tmp/example.py
                  e.__traceback__.tb_lineno  # 2
              )
        
        # Processing a cancel request
        if request_type == 'cancel':
          try:
            employee_id = input('Employee ID > ').split(' ')[0]
            meeting_id = input('Meeting ID > ').split(' ')[0]
            meeting_system.cancel_room(int(employee_id), int(meeting_id))
          except Exception as e:
            print('Error while room cancelation. try again ...',e)
            print(
                  type(e).__name__,          # TypeError
                  __file__,                  # /tmp/example.py
                  e.__traceback__.tb_lineno  # 2
              )

        print('Current schedules - \n', meeting_system.schedules)
        print('Currently scheduled meetings - \n', meeting_system.meetings)
    except Exception as e:
        valid_meeting_initializer = False
        print('Error occurred. Try again ...', e)


CreateMeetingSystem()

.github/scripts/versionCheck.js
/**
 * Runs when using `npm version <type>`
 * Checks changelog added and on latest commit compared to origin
 * Moves unreleased section to new version section
 */
import fs from 'fs';
import moment from 'moment';
import { execSync } from 'child_process';
import { version } from '../../../node-app-setup/package.json';

function check() {
  const branch = execSync('git rev-parse --abbrev-ref HEAD').toString().trim();
  const isPreRelease = /^[\d.]+-.*$/.test(version);
  const isRelease = /^[\d.]+$/.test(version);
  if (!isRelease && !isPreRelease) {
    throw new Error(`Invalid Version Tag: ${version}`);
  }
  if (isRelease && branch !== 'master') {
    throw new Error(
      `\n------------------------------------------------------\
      \nRelease tag can only be made on "master" branch.\
      \nPlease revert change in package.json\
      \n------------------------------------------------------\n`,
    );
  }
  if (isPreRelease && branch !== 'dev') {
    throw new Error(
      `\n------------------------------------------------------\
      \nPre-release tag can only be made on "dev" branch.\
      \nPlease revert change in package.json\
      \n------------------------------------------------------\n`,
    );
  }

  // Updating remote
  execSync('git remote update');

  const status = execSync('git status -uno').toString().trim();
  const isBehind = status.match(/Your branch is behind .+ by (.+) commit.+/gim);
  const isDiff = status.match(
    /Your branch and .+ have diverged,[\s\n\r]+and have (.+) and (.+) different commits each.+/gim,
  );

  if (isBehind) {
    throw new Error(
      `\n------------------------------------------------------\
      \n${isBehind[0]}\
      \nPlease revert change in package.json\
      \n------------------------------------------------------\n`,
    );
  }

  if (isDiff) {
    throw new Error(
      `\n------------------------------------------------------\
      \n${isDiff[0]}\
      \nPlease revert change in package.json\
      \n------------------------------------------------------\n`,
    );
  }

  const UNRELEASED = '[Unreleased]\n\n> NOTE: ⬇⬇ Add below this ⬇⬇';
  const UNRELEASED_END = '\n> NOTE: ⬆⬆ And above this ⬆⬆';

  const changelogFile = fs.readFileSync('CHANGELOG.md').toString();
  const changelogStart = changelogFile.indexOf(UNRELEASED) + UNRELEASED.length;
  const changelogEnd = changelogFile.indexOf(UNRELEASED_END, changelogStart);
  const changelog = changelogFile.slice(changelogStart, changelogEnd).trim();

  if (!changelog.length) {
    throw new Error(
      `\n------------------------------------------------------\
      \nPlease add a changelog entry\
      \nPlease revert change in package.json\
      \n------------------------------------------------------\n`,
    );
  }

  const deployNotesSection = changelogFile.indexOf('## Deploy Notes');
  const deployNotesStart =
    changelogFile.indexOf(UNRELEASED, deployNotesSection) + UNRELEASED.length;
  const deployNotesEnd = changelogFile.indexOf(UNRELEASED_END, deployNotesStart);
  const deployNotes = changelogFile.slice(deployNotesStart, deployNotesEnd).trim();

  console.info(
    `\n------------------CHANGELOG-------------------------\
    \n${changelog}
    \n------------------------------------------------------\n`,
    `\n----------------DEPLOY--NOTES------------------------\
    \n${deployNotes}
    \n------------------------------------------------------\n`,
  );

  if (isPreRelease) return;

  const UNRELEASED_LINK = `---\n\n[unreleased]`;
  const unreleasedLinkStart = changelogFile.indexOf(UNRELEASED_LINK) + UNRELEASED_LINK.length;
  const unreleasedLinkEnd = changelogFile.indexOf('\n', unreleasedLinkStart) + 1;
  const unreleasedLink = changelogFile.slice(unreleasedLinkStart, unreleasedLinkEnd);
  const prevVersion = unreleasedLink.match(/v([\d.]+)\.\.\./)?.[1];

  let updatedChangelog = `${changelogFile.slice(
    0,
    changelogStart,
  )}\n${UNRELEASED_END}\n\n## [${version}] - ${moment().format(
    'YYYY-MM-DD',
  )}\n\n${changelog}${changelogFile.slice(changelogEnd + UNRELEASED_END.length, deployNotesStart)}`;

  if (deployNotes.length) {
    updatedChangelog += `\n${UNRELEASED_END}\n\n### [${version}]\n\n${deployNotes}${changelogFile.slice(
      deployNotesEnd + UNRELEASED_END.length,
      unreleasedLinkStart,
    )}`;
  } else {
    updatedChangelog += changelogFile.slice(deployNotesStart, unreleasedLinkStart);
  }

  const versionCompare = (from, to) =>
    `https://github.com/dbads/node-app-setup/compare/v${from}...${to ? `v${to}` : 'dev'}`;

  updatedChangelog += `: ${versionCompare(version)}\n[${version}]: ${versionCompare(
    prevVersion,
    version,
  )}\n${changelogFile.slice(unreleasedLinkEnd)}`;

  fs.writeFileSync('CHANGELOG.md', updatedChangelog);
  execSync('git add CHANGELOG.md');
}

if (require.main?.filename === __filename) {
  check();
}


sample_big_module.py
# Notes
# 1. the meeting system is designed to handle operatinos within a single day
# 2. in a day time is measured using 24 hour clock 1,2,3...24
# 3. time inputs are allowed only as [1,2 .... 23, 24] discrete values


class MeetingSystem:
  """ A model to create a meeting system with following methods
  - book: input format = book employee_id, start_time, end_time
  - cancel: input format = cancel employee_id, meeting_id

  Helper/utitlity methods
  - get_time_key: returns a string key for given time interval
  - confirm_room: books a room with gicen details
  """
  schedules = {
  # sample entry
  # '12': [5,7], i.e room_id 5 and 7 is busy during time 1 to 2
  }

  meetings = {
    # An sample entry
    # meeting_id: {
    #   'employee_id': demo_employee_id,
    #   'start_time': demo_start_time,
    #   'end_time': demo_end_time,
    #   'room_id': demo_room_id,
    # }
  }

  rooms_count = 0
  employees_count = 0

  # Initializing meetings system
  def __init__(self, rooms_count, employees_count):
    """ Just initiallizes the meeting system with employees_count and rooms_count
    """
    self.rooms_count = rooms_count
    self.employees_count = employees_count


  # Utitlity methods
  def validate_meeting_details(self, details):
    """Checks if meeting details are valid
    - time should be in [1,24]
    - employee_id should be in [1,N], N is employees_count
    - meeting_id should exist in meetings
    - @return Boolean(True/False)
    """
    is_valid = True

    # time should be a discrete value in [1,24], end_time>start_time
    if 'start_time' in details and 'end_time' in details:
      start_time = details['start_time']
      end_time = details['end_time']
      if not type(start_time) == int or not type(end_time) == int or\
        start_time not in range(1 ,25) or end_time not in range(1, 25):
        print('Invalid time provided.')
        is_valid = False

      if start_time >= end_time:
        print('start_time should be strictly less than end_time.')
        is_valid = False

      if end_time - start_time > 3:
        print('Meeting should not exceed 3 hours.')
        is_valid = False

    # check if employee_id is valid
    if 'employee_id' in details:
      employee_id = details['employee_id']
      if type(employee_id) != int or employee_id not in range(1, self.employees_count + 1):
        print('Invalid Employee ID.')
        is_valid = False

    # validate meeting_id
    if 'meeting_id' in details:
      meeting_id = details['meeting_id']
      if meeting_id not in self.meetings:
        print('Invalid meeting_id.')
        is_valid = False

    if not is_valid: return False
    return True


  def get_time_key(self, start_time, end_time):
    """ Helper to get string key of time interval
    - @input start_time and end_time
    - @returns string key - for this time interval
    """
    return str(start_time) + '-' + str(end_time)


  def confirm_room(self, employee_id, start_time, end_time):
    """Schedule a meeting in the first room which is free at the requested time slot
    - @input employee_id, start_time, end_time
    - @returns void
    """
    time_key = self.get_time_key(start_time, end_time)

    # schedule a meeting in free_room_id
    # making room busy for this time interval
    if time_key not in self.schedules:
      free_room_id = 1
      self.schedules[time_key] = [free_room_id] # no room engaged this time, simple take room_id 1 and schedule meeting there
    else:
      engaged_rooms = self.schedules[time_key]
      free_room_id = len(engaged_rooms) + 1
      self.schedules[time_key].append(free_room_id)

    # create a meeting at this time
    self.meetings[len(self.meetings) + 1] = {
      'room_id': free_room_id,
      'employee_id': employee_id,
      'start_time': start_time,
      'end_time': end_time,
    }
    print('Successfuly booked room for meeting')


  def is_overlaping_time(self, start_time, end_time, new_start_time, new_end_time):
    """Checks whether time intervals overlap or not
    """
    return new_start_time in range(start_time, end_time+1) or new_end_time in range(start_time, end_time+1)


  # core methods
  def book_room(self, employee_id, start_time, end_time):
    """Schedules a meeting in a room with requested details
    """
    is_valid_details = self.validate_meeting_details({'employee_id': employee_id, 'start_time': start_time, 'end_time': end_time})

    if is_valid_details:
      time_key = self.get_time_key(start_time, end_time)

      # see if the employee has already a meeting at this time
      already_booked_meeting = False # suppose not booked a meeting already by this member
      # see if one of the room is having a meeting scheduled by employee_id
      for room_id, meeting in self.meetings.items():
        if meeting['employee_id'] == employee_id\
          and self.is_overlaping_time(meeting['start_time'], meeting['end_time'], start_time, end_time):
          already_booked_meeting = True # found a meeting booked by this employee_id at same time
          print('Employee has already a meeting scheduled at same time')

      if not already_booked_meeting:
        # even though this employee has not booked a meeting at this time, there are already other meetings going on at this time
        # see if there is a free room to schedule this meeting at this time
        engaged_rooms = self.schedules[time_key] if time_key in self.schedules else []
        if len(engaged_rooms) + 1 > self.rooms_count: # all rooms are already engaged for this time
          print('All rooms busy for the given time interval')
        else:
          self.confirm_room(employee_id, start_time, end_time)


  def cancel_room(self, employee_id, meeting_id):
    # validate input
    is_valid_details = self.validate_meeting_details({'employee_id': employee_id, 'meeting_id': meeting_id})

    if is_valid_details:
      # cancel the meeting if employee_id has created this meeting
      if self.meetings[meeting_id]['employee_id'] == employee_id:
        start_time = self.meetings[meeting_id]['start_time']
        end_time = self.meetings[meeting_id]['end_time']
        time_key = self.get_time_key(start_time, end_time)
        room_id = self.meetings[meeting_id]['room_id']
        self.schedules[time_key].remove(room_id) # free the room from schedules
        del self.meetings[meeting_id] # remove meeting detail from meetings
        print('Successfuly canceled the meeting')
      else: print('You are not the organizer of this meeting')


def CreateMeetingSystem():
  valid_meeting_initializer = False
  while not valid_meeting_initializer:
    employees_count = input('Please Enter the no of employees > ')
    rooms_count = input('Please Enter the no of rooms > ')

    try:
      employees_count = int(employees_count)
      rooms_count = int(rooms_count)
      valid_meeting_initializer = True # valid a meeting initiallizers rooms_count and employees_count, stop calling initializer again now

      meeting_system = MeetingSystem(rooms_count, employees_count)
      print('A MeetingSystem with %s Employees and %s meeting rooms has been created.\n'%(employees_count, rooms_count))

      print('Notes :-\n')
      print('1. Employee IDs should be in [1-N], N is the employee_count')
      print('2. Time should be in [1,24]\n')

      request_type = 1
      while request_type != 'quit':
        request_type = input('Enter the request_type [book, cancel, quit] > ')
        if request_type not in ['book', 'cancel', 'quit']:
          print('Operations not allowed. Please try again ...')

        # processing book Operations
        if request_type == 'book':
          try:
            employee_id = input('Employee ID > ').split(' ')[0]
            start_time = input('Start time > ').split(' ')[0]
            end_time = input('End time > ').split(' ')[0]
            meeting_system.book_room(int(employee_id), int(start_time), int(end_time))
          except Exception as e:
            print('Error while booking room. try again ...', e)
            print(
                  type(e).__name__,          # TypeError
                  __file__,                  # /tmp/example.py
                  e.__traceback__.tb_lineno  # 2
              )

        # Processing a cancel request
        if request_type == 'cancel':
          try:
            employee_id = input('Employee ID > ').split(' ')[0]
            meeting_id = input('Meeting ID > ').split(' ')[0]
            meeting_system.cancel_room(int(employee_id), int(meeting_id))
          except Exception as e:
            print('Error while room cancelation. try again ...',e)
            print(
                  type(e).__name__,          # TypeError
                  __file__,                  # /tmp/example.py
                  e.__traceback__.tb_lineno  # 2
              )

        print('Current schedules - \n', meeting_system.schedules)
        print('Currently scheduled meetings - \n', meeting_system.meetings)
    except Exception as e:
        valid_meeting_initializer = False
        print('Error occurred. Try again ...', e)


CreateMeetingSystem()